File: common/src/main/java/cn/breadnicecat/candycraftce/block/blocks/SensitiveJellyBlock.java
package cn.breadnicecat.candycraftce.block.blocks;

import cn.breadnicecat.candycraftce.block.CBlockTags;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.redstone.Redstone;
import net.minecraft.world.phys.AABB;
import org.jetbrains.annotations.NotNull;

/**
 * Created in 2024/2/14 10:54
 * Project: candycraftce
 *
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 * <p>
 */
public class SensitiveJellyBlock extends JellyBlock {
	
	public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
	public static final JellyType SENSITIVE = new JellyType(0.0, 1F);
	
	public SensitiveJellyBlock(Properties properties, JellyType type) {
		super(properties, type);
		registerDefaultState(stateDefinition.any().setValue(POWERED, false));
	}
	
	public SensitiveJellyBlock(Properties properties) {
		this(properties, SENSITIVE);
	}
	
	@SuppressWarnings("deprecation")
	@Override
	public void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston) {
		level.scheduleTick(pos, this, 1);
	}
	
	@SuppressWarnings("deprecation")
	@Override
	public void tick(BlockState state, ServerLevel level, final BlockPos pos, RandomSource random) {
		BlockPos detect = pos;
		BlockPos above = pos.above();
		BlockState state1 = level.getBlockState(above);
		if (state1.is(CBlockTags.BT_JELLY.b()) && !state1.is(this)) {
			detect = above;
		}
		boolean powered = !level.getEntitiesOfClass(LivingEntity.class, new AABB(
				detect.getX(),
				detect.getY() + 1,
				detect.getZ(),
				detect.getX() + 1,
				detect.getY() + 1.5,
				detect.getZ() + 1
		)).isEmpty();
		
		if (powered != state.getValue(POWERED)) level.setBlockAndUpdate(pos, state.setValue(POWERED, powered));
		level.scheduleTick(pos, this, 1);
	}
	
	@Override
	protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
		builder.add(POWERED);
		super.createBlockStateDefinition(builder);
	}
	
	@SuppressWarnings("deprecation")
	@Override
	public boolean isSignalSource(@NotNull BlockState state) {
		return true;
	}
	
	@SuppressWarnings("deprecation")
	@Override
	public int getSignal(BlockState state, @NotNull BlockGetter level, @NotNull BlockPos pos, @NotNull Direction direction) {
		return state.getValue(POWERED) ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_NONE;
	}
	
	
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/core/block/blocks/jelly/SensitiveJellyBlock.kt
package cn.breadnicecat.candycraftce.core.block.blocks.jelly

import cn.breadnicecat.candycraftce.core.tag.CTags
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.RandomSource
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.level.block.state.properties.BooleanProperty
import net.minecraft.world.level.redstone.Redstone
import net.minecraft.world.phys.AABB

/**
 * Created by NiceCat on 2026/2/27.
 * Project: candycraftce
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 *
 */
class SensitiveJellyBlock(properties: Properties) : JellyBlock(properties) {
    companion object {
        val POWERED: BooleanProperty = BlockStateProperties.POWERED
    }

    init {
        registerDefaultState(stateDefinition.any().setValue(POWERED, false))
    }

    @Deprecated("Deprecated in Java")
    override fun onPlace(state: BlockState, level: Level, pos: BlockPos, oldState: BlockState, movedByPiston: Boolean) {
        level.scheduleTick(pos, this, 1)
    }

    @Deprecated("Deprecated in Java")
    override fun tick(state: BlockState, level: ServerLevel, pos: BlockPos, random: RandomSource) {
        var detectPos: BlockPos = pos
        val above = pos.above()
        val aboveState = level.getBlockState(above)
        if (aboveState.`is`(CTags.jelly.block) && !aboveState.`is`(this)) {
            detectPos = above
        }
        val powered = !level.getEntitiesOfClass(
            LivingEntity::class.java, AABB(
                detectPos.x.toDouble(),
                (detectPos.y + 1).toDouble(),
                detectPos.z.toDouble(),
                (detectPos.x + 1).toDouble(),
                detectPos.y + 1.5,
                (detectPos.z + 1).toDouble()
            )
        ).isEmpty()

        if (powered != state.getValue(POWERED)) {
            level.setBlockAndUpdate(pos, state.setValue(POWERED, powered))
        }
        level.scheduleTick(pos, this, 1)
    }

    protected override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block?, BlockState?>) {
        builder.add(POWERED)
        super.createBlockStateDefinition(builder)
    }

    @Deprecated("Deprecated in Java")
    override fun isSignalSource(state: BlockState) = true

    @Deprecated("Deprecated in Java")
    override fun getSignal(state: BlockState, level: BlockGetter, pos: BlockPos, direction: Direction): Int {
        return if (state.getValue(POWERED)) Redstone.SIGNAL_MAX else Redstone.SIGNAL_NONE
    }


}